Add methods to convert CMSampleBuffer and CMFormatDescription to MediaSamplesBlock and TrackInfo

jyavenard opened the pull request:
#### 587602e420de43c48d5fd31e7095ce4dec1dc244
<pre>
Add methods to convert CMSampleBuffer and CMFormatDescription to MediaSamplesBlock and TrackInfo
<a href="https://bugs.webkit.org/show_bug.cgi?id=284701">https://bugs.webkit.org/show_bug.cgi?id=284701</a>
<a href="https://rdar.apple.com/141498917">rdar://141498917</a>

Reviewed by Youenn Fablet.

Add createAudioInfoFromFormatDescription and samplesBlockFromCMSampleBuffer.
MediaSamplesBlock created share the CMSampleBuffer&apos;s CMBlockBuffer, no new memory is allocated.

We remove support for the UseByteRange variant in MediaSamplesBlock and SourceBufferParserWebM
as it was only usd by the MediaFormatReader plugin which got removed in 282704@main.

New methods aren&apos;t used yet and will be used in a follow-up change.
No change in behaviours, covered by existing tests.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSample.h:
Make MediaSamplesBlock use FastMalloc so that it can be allocated by makeUniqueRef.
Change MediaSampleItem&apos;s trimDuration to trimInterval so that it allows 1:1
compatible with CMSampleBuffer&apos;s kCMSampleBufferAttachmentKey_TrimDurationAtEnd.
Remote unused profile and extendedProfile members; we use the FourCC codecName instead.
Change flags type to uint32_t to prepare for IPC serialisation.
Add various utility methods to replicate MediaSample ones (presentationTime, presentationEndTime, duration, isSync).
Sured RefPtr&lt;SharedBuffer&gt; instead of RefPtr&lt;const SharedBuffer&gt; as the IPC serialisation
doesn&apos;t handle const type properly.

(WebCore::MediaSamplesBlock::MediaSampleItem::isSync const):
(WebCore::MediaSamplesBlock::MediaSamplesBlock):
(WebCore::MediaSamplesBlock::presentationTime const):
(WebCore::MediaSamplesBlock::duration const):
(WebCore::MediaSamplesBlock::presentationEndTime const):
(WebCore::MediaSamplesBlock::isSync const):
(WebCore::MediaSamplesBlock::trackID const): Change type to TrackID for consistency.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::audioStreamDescriptionFromAudioInfo):
(WebCore::createAudioFormatDescription):
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createAudioInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
(WebCore::getPacketDescriptions):
(WebCore::ensureContiguousBlockBuffer):
(WebCore::sharedBufferFromCMBlockBuffer): Utility method to create a SharedBuffer from a
CMBlockBuffer. The memory is referenced in place, no copy is performed.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnTrackEntry):
(WebCore::WebMParser::TrackData::readFrameData):
(WebCore::WebMParser::VideoTrackData::consumeFrameData):
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
(WebCore::SourceBufferParserWebM::parsedMediaData):
(WebCore::SourceBufferParserWebM::returnSamples):
(WebCore::SourceBufferParserWebM::flushPendingAudioSamples):
(WebCore::WebMParser::createByteRangeSamples): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
(WebCore::WebMParser::TrackData::createByteRangeSamples): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame): Update method
to use C++20 designated initialisers.

Canonical link: <a href="https://commits.webkit.org/287870@main">https://commits.webkit.org/287870@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd509efb78a2cf246e29f992a05fca7ae393dfd7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81124 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/649 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35067 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85653 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32110 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83234 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/667 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8456 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63333 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21098 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84193 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/419 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73830 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43631 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/316 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27992 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30568 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28552 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87088 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8354 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5917 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71636 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69665 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70871 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17653 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14915 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13847 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8315 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13838 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8152 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11672 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9960 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/MediaSample.h:
```diff
@@ -206,26 +204,46 @@ struct AudioInfo : public TrackInfo {
+    }
+    MediaTime presentationEndTime() const { return presentationTime() + duration(); }
+    bool isSync() const { return size() ? (first().flags & MediaSample::IsSync) : false; }
+    TrackID trackID() const { return m_info ? m_info->trackID : -1; }
```
Might be worth to add a constant for -1, or use a std::optional<TrackID> instead.

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
```diff
@@ -47,9 +50,18 @@ struct PlatformVideoColorSpace;
+WEBCORE_EXPORT Expected<RetainPtr<CMSampleBufferRef>, CString> toCMSampleBuffer(const MediaSamplesBlock&, CMFormatDescriptionRef = nullptr);
+// Convert CMSampleBufferRef to the equivalent MediaSamplesBlock. If TrackInfo
+// is set it will be used, otherwise it will be created from the CMSampleBufferRef's CMFormatDescriptionRef.
+WEBCORE_EXPORT UniqueRef<MediaSamplesBlock> samplesBlockFromCMSampleBuffer(CMSampleBufferRef, TrackInfo* = nullptr);
```
Should probably be const TrackInfo.

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -55,24 +55,34 @@
-static RetainPtr<CMFormatDescriptionRef> createAudioFormatDescription(const AudioInfo& info)
+CAAudioStreamDescription audioStreamDescriptionFromAudioInfo(const AudioInfo& info)
 {
+    ASSERT(info.codecName.value != kAudioFormatLinearPCM);
```
It is not clear to me how we prevent calling `audioStreamDescriptionFromAudioInfo` with a codec that is not PCM.
Maybe we should return an optional which would be null for PCM?

youennf commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1095,7 +1082,12 @@ WebMParser::ConsumeFrameDataResult WebMParser::VideoTrackData::consumeFrameData(
+    m_pendingMediaSamples.append({
+        .presentationTime = presentationTime,
+        .decodeTime = presentationTime,
+        .data = WTFMove(m_completeFrameData),
```
We usually prefer using std::exchange.

youennf commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1310,7 +1302,13 @@ WebMParser::ConsumeFrameDataResult WebMParser::AudioTrackData::consumeFrameData(
+        .presentationTime = localPresentationTime,
+        .duration = packetDuration,
+        .trimInterval = { trimDuration, MediaTime::zeroTime() },
+        .data = WTFMove(m_completeFrameData),
```
Ditto.

jyavenard committed:
Add methods to convert CMSampleBuffer and CMFormatDescription to MediaSamplesBlock and TrackInfo
https://bugs.webkit.org/show_bug.cgi?id=284701
rdar://141498917

Reviewed by Youenn Fablet.

Add createAudioInfoFromFormatDescription and samplesBlockFromCMSampleBuffer.
MediaSamplesBlock created share the CMSampleBuffer's CMBlockBuffer, no new memory is allocated.

We remove support for the UseByteRange variant in MediaSamplesBlock and SourceBufferParserWebM
as it was only usd by the MediaFormatReader plugin which got removed in 282704@main.

New methods aren't used yet and will be used in a follow-up change.
No change in behaviours, covered by existing tests.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSample.h:
Make MediaSamplesBlock use FastMalloc so that it can be allocated by makeUniqueRef.
Change MediaSampleItem's trimDuration to trimInterval so that it allows 1:1
compatible with CMSampleBuffer's kCMSampleBufferAttachmentKey_TrimDurationAtEnd.
Remote unused profile and extendedProfile members; we use the FourCC codecName instead.
Change flags type to uint32_t to prepare for IPC serialisation.
Add various utility methods to replicate MediaSample ones (presentationTime, presentationEndTime, duration, isSync).
Sured RefPtr<SharedBuffer> instead of RefPtr<const SharedBuffer> as the IPC serialisation
doesn't handle const type properly.

(WebCore::MediaSamplesBlock::MediaSampleItem::isSync const):
(WebCore::MediaSamplesBlock::MediaSamplesBlock):
(WebCore::MediaSamplesBlock::presentationTime const):
(WebCore::MediaSamplesBlock::duration const):
(WebCore::MediaSamplesBlock::presentationEndTime const):
(WebCore::MediaSamplesBlock::isSync const):
(WebCore::MediaSamplesBlock::trackID const): Change type to TrackID for consistency.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::audioStreamDescriptionFromAudioInfo):
(WebCore::createAudioFormatDescription):
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createAudioInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
(WebCore::getPacketDescriptions):
(WebCore::ensureContiguousBlockBuffer):
(WebCore::sharedBufferFromCMBlockBuffer): Utility method to create a SharedBuffer from a
CMBlockBuffer. The memory is referenced in place, no copy is performed.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnTrackEntry):
(WebCore::WebMParser::TrackData::readFrameData):
(WebCore::WebMParser::VideoTrackData::consumeFrameData):
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
(WebCore::SourceBufferParserWebM::parsedMediaData):
(WebCore::SourceBufferParserWebM::returnSamples):
(WebCore::SourceBufferParserWebM::flushPendingAudioSamples):
(WebCore::WebMParser::createByteRangeSamples): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
(WebCore::WebMParser::TrackData::createByteRangeSamples): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame): Update method
to use C++20 designated initialisers.

Canonical link: https://commits.webkit.org/287870@main
